Reviving parallel exception code for PETSc 3.6

jwpeterson opened the pull request:
This branch revives @friedmud's exception throwing code and expands on it a bit. It works for me on my laptop using PETSc-3.6, I'm putting it up here to see how the build system's PETSc deals with it. The new tests don't work for PETSc < 3.6.0 so they are currently restricted.  I think it's still worth getting the code in to MOOSE sooner rather than later.

This branch requires a patch in BISON, there is already a merge request in place for that.

Refs #3777

friedmud committed:
Parallel exception handling.

The basic idea is this:

1.  Exceptions are thrown _and_ caught within threaded regions (an exception never traverses a threaded call).
2.  When it's caught within the threaded region a flag is set in MOOSE.
3.  Other threads can check that flag to see if they should quit early (or they can keep working - who cares).
4.  Once the threaded region is finished the flag is then checked and communicated globally over MPI.
5.  If there was an exception PETSc is informed and the rest of the operation (like residual computation) is (possibly) aborted.
6.  PETSc should stop the current solve (it's currently not)
6.  The Executioner should get the flag back from PETSc that the solve failed and do what it normally does when the solve fails (like cut timesteps).

refs #3777

Rebased from Aug 2014 MOOSE to June 2015 MOOSE by John W. Peterson <jwpeterson@gmail.com>.

Conflicts:
framework/include/base/ThreadedElementLoopBase.h
framework/src/base/AuxiliarySystem.C
framework/src/base/FEProblem.C
framework/src/base/NonlinearSystem.C
test/src/executioners/ExceptionSteady.C
test/tests/misc/exception/parallel_exception_residual.i

jwpeterson committed:
Misc. commits while working on Exception:

* It is no longer necessary to set type=Exception in the exception
  tests, the Exception is thrown because of the ExceptionKernel not
  the Excecutioner.

* FEProblem::setException() should take a const reference.

* We don't want to call checkExceptionAndStopSolve() after computing the
  initial residual because it throws an exception.  We need to trust
  that the exception has already been handled at a lower level and PETSc
  will return a diverged flag during it's next reduction.

* Throw on call number 1000 in ExceptionKernel.

* parallel_exception_residual.i doesn't need -pc_factor_shift_type if
  we put the Kernels on all blocks.

* More specific message from ExceptionKernel::computeQpResidual()

* Catch and don't re-throw in NonlinearSystem::computeJacobian()

* Modify the ExceptionKernel to only throw after a specific number of
  calls to computeQpJacobian() so we don't throw during the routine
  where the sparsity of the preconditioner is constructed.

* Throw an exception from ExceptionKernel rather than calling stopSolve().

* Set NaN in the residual vector during stopSolve().

* Adding transient versions of tests which throw from the residual and Jacobian.

* Once checkExceptionAndStopSolve() has handled the exception, set
  _has_exception to false.

* In the ExceptionKernel, increment the call number before we throw,
  so the Kernel can continue to be used after that point instead of
  throwing exceptions every time it is called subsequently.

* Remove the original exception_test, which was already deleted.

* Removing original exception residual and jacobian tests.
  They are redundant with the transient version of these tests.

* Changing the parallel_exception_initial_condition test to RunApp
  We are testing that we get the output string:

  'MooseException thrown during initial condition computation'

  This test is somewhat independent of the the residual test, since the
  exception is thrown from a different invocation of computeResidual().

* Add throw() specification to the MooseException destructor.  This
  declares that it won't throw an exception, and was required for GCC.

* The AdaptiveTransient Exectioner throws a MooseException, so it
  needs to #include "MooseException.h"

* Restrict new tests to only run on PETSc >= 3.6.0

  Otherwise, the error message from the residual test is:

  ```
  Time Step  0, time = 0
                  dt = 0

  Time Step  1, time = 0.01
                  dt = 0.01
   0 Nonlinear |R| = 2.449490e+00
        0 Linear |R| = 2.449490e+00
        1 Linear |R| = 1.133359e-02
  MooseException thrown during residual calculation
        2 Linear |R| = nan
  [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
  [0]PETSC ERROR: Petsc has generated inconsistent data
  [0]PETSC ERROR: Computed Nan differencing parameter h
  [0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
  [0]PETSC ERROR: Petsc Release Version 3.5.2, Sep, 08, 2014
  ```

  In the Jacobian test, the NaN we insert is not even enough to
  trigger a PETSc error in older PETScs, the example just continues to
  run with an incorrect Jacobian.

Refs #3777

moosebuild commented:
Results of testing 12f9c45f27cad8fd0d47fe3806e298179de5b05b using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17273

moosebuild commented:
Results of testing 12f9c45f27cad8fd0d47fe3806e298179de5b05b using moose_PR_app_tests recipe:

Failed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17276

moosebuild commented:
Results of testing 12f9c45f27cad8fd0d47fe3806e298179de5b05b using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17274

moosebuild commented:
Results of testing 12f9c45f27cad8fd0d47fe3806e298179de5b05b using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17275

moosebuild commented on framework/include/base/AuxiliarySystem.h:
```diff
@@ -130,7 +130,7 @@ class AuxiliarySystem : public SystemTempl<TransientExplicitSystem>
   void computeElementalVars(ExecFlagType type);
 
-  FEProblem & _mproblem;
+  FEProblem & _fe_problem;
```
lol!

moosebuild commented on framework/include/base/FEProblem.h:
```diff
@@ -1022,6 +1046,12 @@ class FEProblem :
   std::string _pc_description;
 
+  /// Whether or not an exception has occurred
+  bool _has_exception;
```
These should both really be private.  Not a big deal... but we do need to get better about making things private (Yes, @permcody , I'm agreeing with you!)

friedmud replied:
I disagree with @moosebuild ... I think this is fine where it is.

permcody replied:
@moosebuild - you definitely have your :hankey: together. More private is best!

friedmud commented on framework/src/base/FEProblem.C:
```diff
@@ -3837,6 +3875,9 @@ FEProblem::checkNonlinearConvergence(std::string &msg,
                                      const Real initial_residual_before_preset_bcs,
                                      const Real div_threshold)
 {
+  if (_has_exception)
```
This won't really work, right?  Because you're clearing `_has_exception` when you handle it.

I think you just need to check the PETSc reason and if it's `DIVERGED_DOMAIN_ERROR` then set this.

jwpeterson replied:
I agree that this doesn't make sense, but at this point, we are actually *computing* the reason, so I don't think we can check the reason.  Instead, we could call `SNESGetFunctionDomainError` (returns true if `SNESSetFunctionDomainError()` has been called) in our `petscNonlinearConverged()` function, and set the appropriate reason there.  Not sure that's actually necessary, as PETSc already seems to diverge on its own with this reason, but I will test it.

friedmud commented on framework/src/base/FEProblem.C:
```diff
@@ -3907,6 +3948,9 @@ FEProblem::checkLinearConvergence(std::string & /*msg*/,
                                   const Real /*dtol*/,
                                   const PetscInt maxits)
 {
+  if (_has_exception)
```
Same here

friedmud commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -464,6 +480,9 @@ NonlinearSystem::setupSplitBasedPreconditioner()
 bool
 NonlinearSystem::converged()
 {
+  if (_fe_problem.hasException())
```
Again, this won't work...

jwpeterson committed:
_has_exception is now cleared before the convergence check, so this code branch is no longer reached.

jwpeterson committed:
Call SNESGetFunctionDomainError() from petscNonlinearConverged()

Set SNES_DIVERGED_FUNCTION_DOMAIN and return, if necessary.  Not sure
we will reach this code, as PETSc seems to diverge with this reason on
its own already...

moosebuild commented:
Results of testing 407e6f847e6cbf40cc6faaa42ffb96fa402ae56b using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17299

moosebuild commented:
Results of testing 407e6f847e6cbf40cc6faaa42ffb96fa402ae56b using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17300

moosebuild commented:
Results of testing 407e6f847e6cbf40cc6faaa42ffb96fa402ae56b using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17302

moosebuild commented:
Results of testing 407e6f847e6cbf40cc6faaa42ffb96fa402ae56b using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.org/view_job/17301
